File: governance/engine/orchestrator/state_machine.py
"""Phase state machine with gate enforcement at every boundary.

The state machine is the core of the orchestrator. It enforces:
1. Only valid phase transitions are allowed
2. Every transition passes through a capacity gate check
3. Orange/Red tiers raise ShutdownRequired (cannot be bypassed)
4. All gate decisions are recorded for audit
"""

from __future__ import annotations

from dataclasses import dataclass, field

from governance.engine.orchestrator.capacity import (
    Action,
    CapacitySignals,
    Tier,
    classify_tier,
    gate_action,
)


class InvalidTransition(Exception):
    """Raised when an invalid phase transition is attempted."""

    def __init__(self, from_phase: int, to_phase: int):
        self.from_phase = from_phase
        self.to_phase = to_phase
        super().__init__(
            f"Invalid transition: Phase {from_phase} → Phase {to_phase}"
        )


class ShutdownRequired(Exception):
    """Raised when a gate check requires shutdown.

    Callers must treat this as fatal and must not continue the loop
    after catching it. The only valid response is to execute the
    shutdown protocol (write checkpoint, clean git state, exit).
    """

    def __init__(
        self,
        tier: Tier,
        action: Action,
        gates_passed: list[dict],
        signals: CapacitySignals | None = None,
    ):
        self.tier = tier
        self.action = action
        self.gates_passed = gates_passed
        self.signals = signals
        super().__init__(
            f"Shutdown required: tier={tier.value}, action={action.value}"
        )


# ---------------------------------------------------------------------------
# Valid phase transitions — extracted from startup.md pipeline overview
# ---------------------------------------------------------------------------

VALID_TRANSITIONS: dict[int, frozenset[int]] = {
    0: frozenset({1, 2, 3, 4, 5, 6, 7}),  # Phase 0 (recovery) can resume to any phase
    1: frozenset({2}),                      # Pre-flight → Planning
    2: frozenset({3}),                      # Planning → Dispatch
    3: frozenset({4}),                      # Dispatch → Review
    4: frozenset({3, 5}),                   # Review → Re-dispatch (feedback) or Merge
    5: frozenset({1, 6}),                   # Merge → Loop or Build (deployment)
    6: frozenset({7}),                      # Build → Deploy
    7: frozenset({1}),                      # Deploy → Loop back to Pre-flight
}

# Actions that require shutdown
_SHUTDOWN_ACTIONS = frozenset({Action.EMERGENCY_STOP, Action.CHECKPOINT})


@dataclass
class PhaseState:
    """Current pipeline phase."""

    phase: int = 0
    sub_phase: str | None = None


@dataclass
class GateRecord:
    """Record of a gate check for audit trail."""

    phase: int
    tier: str
    action: str
    tool_calls: int
    turns: int
    issues_completed: int


class StateMachine:
    """Deterministic phase state machine with gate enforcement.

    Every transition passes through a capacity gate. If the gate
    determines the tier is Orange or Red (for most phases), it raises
    ShutdownRequired. The caller cannot proceed.
    """

    def __init__(self, parallel_coders: int = 5):
        self.state = PhaseState(phase=0)
        self.signals = CapacitySignals(parallel_coders=parallel_coders)
        self.gates_passed: list[GateRecord] = []
        self._started = False

    @property
    def phase(self) -> int:
        return self.state.phase

    @property
    def tier(self) -> Tier:
        return classify_tier(self.signals)

    def transition(self, target_phase: int) -> Action:
        """Attempt a phase transition with gate enforcement.

        Args:
            target_phase: The phase to transition to (0-5).

        Returns:
            The action the orchestrator should take (PROCEED, SKIP_DISPATCH,
            or FINISH_CURRENT).

        Raises:
            InvalidTransition: If the transition is not valid from current phase.
            ShutdownRequired: If the gate check requires shutdown.
        """
        # First transition from Phase 0 is always valid (startup)
        if not self._started:
            self._started = True
        else:
            if target_phase not in VALID_TRANSITIONS.get(self.state.phase, frozenset()):
                raise InvalidTransition(self.state.phase, target_phase)

        # Gate check
        tier = classify_tier(self.signals)
        action = gate_action(target_phase, tier)

        # Record gate decision
        record = GateRecord(
            phase=target_phase,
            tier=tier.value,
            action=action.value,
            tool_calls=self.signals.tool_calls,
            turns=self.signals.turns,
            issues_completed=self.signals.issues_completed,
        )
        self.gates_passed.append(record)

        # Enforce shutdown
        if action in _SHUTDOWN_ACTIONS:
            raise ShutdownRequired(
                tier=tier,
                action=action,
                gates_passed=[
                    {
                        "phase": g.phase,
                        "tier": g.tier,
                        "action": g.action,
                    }
                    for g in self.gates_passed
                ],
                signals=self.signals,
            )

        # Transition
        self.state = PhaseState(phase=target_phase)
        return action

    def record_tool_call(self) -> Tier:
        """Increment tool call counter and return current tier."""
        self.signals.tool_calls += 1
        return classify_tier(self.signals)

    def record_turn(self) -> Tier:
        """Increment turn counter and return current tier."""
        self.signals.turns += 1
        return classify_tier(self.signals)

    def record_issue_completed(self) -> Tier:
        """Increment issues completed and return current tier."""
        self.signals.issues_completed += 1
        return classify_tier(self.signals)

    def get_gate_history(self) -> list[dict]:
        """Return gate history for checkpoint context_gates_passed field."""
        return [
            {"phase": g.phase, "tier": g.tier, "action": g.action}
            for g in self.gates_passed
        ]

    def to_dict(self) -> dict:
        """Serialize state machine for persistence across CLI invocations."""
        return {
            "phase": self.state.phase,
            "sub_phase": self.state.sub_phase,
            "signals": {
                "tool_calls": self.signals.tool_calls,
                "turns": self.signals.turns,
                "issues_completed": self.signals.issues_completed,
                "parallel_coders": self.signals.parallel_coders,
                "system_warning": self.signals.system_warning,
                "degraded_recall": self.signals.degraded_recall,
            },
            "gates_passed": [
                {
                    "phase": g.phase,
                    "tier": g.tier,
                    "action": g.action,
                    "tool_calls": g.tool_calls,
                    "turns": g.turns,
                    "issues_completed": g.issues_completed,
                }
                for g in self.gates_passed
            ],
            "started": self._started,
        }

    @classmethod
    def from_dict(cls, data: dict) -> StateMachine:
        """Restore state machine from persisted dict."""
        signals = data.get("signals", {})
        sm = cls(parallel_coders=signals.get("parallel_coders", 5))
        sm.state = PhaseState(
            phase=data.get("phase", 0),
            sub_phase=data.get("sub_phase"),
        )
        sm.signals.tool_calls = signals.get("tool_calls", 0)
        sm.signals.turns = signals.get("turns", 0)
        sm.signals.issues_completed = signals.get("issues_completed", 0)
        sm.signals.system_warning = signals.get("system_warning", False)
        sm.signals.degraded_recall = signals.get("degraded_recall", False)
        sm.gates_passed = [
            GateRecord(
                phase=g["phase"],
                tier=g["tier"],
                action=g["action"],
                tool_calls=g.get("tool_calls", 0),
                turns=g.get("turns", 0),
                issues_completed=g.get("issues_completed", 0),
            )
            for g in data.get("gates_passed", [])
        ]
        sm._started = data.get("started", False)
        return sm

File: src/internal/orchestrator/statemachine/statemachine.go
// Package statemachine implements the phase state machine for the orchestrator.
// It enforces valid phase transitions, performs capacity gate checks on every
// transition, and records a gate history for auditing.
//
// Ported from Python: governance/engine/orchestrator/state_machine.py
package statemachine

import (
	"fmt"

	"github.com/SET-Apps/ai-submodule/src/internal/orchestrator/capacity"
)

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

// InvalidTransition is returned when a phase transition is not permitted.
type InvalidTransition struct {
	From int
	To   int
}

func (e *InvalidTransition) Error() string {
	return fmt.Sprintf("invalid transition: phase %d → %d", e.From, e.To)
}

// ShutdownRequired is returned when a gate check demands checkpoint or stop.
type ShutdownRequired struct {
	Tier        capacity.Tier
	Action      capacity.Action
	GatesPassed int
	Signals     capacity.Signals
}

func (e *ShutdownRequired) Error() string {
	return fmt.Sprintf(
		"shutdown required: tier=%s action=%s gates_passed=%d",
		e.Tier, e.Action, e.GatesPassed,
	)
}

// ---------------------------------------------------------------------------
// Valid transitions
// ---------------------------------------------------------------------------

// validTransitions maps each phase to its set of legal target phases.
var validTransitions = map[int]map[int]bool{
	0: {1: true, 2: true, 3: true, 4: true, 5: true, 6: true, 7: true},
	1: {2: true},
	2: {3: true},
	3: {4: true},
	4: {3: true, 5: true},
	5: {1: true, 6: true},
	6: {7: true},
	7: {1: true},
}

// ---------------------------------------------------------------------------
// GateRecord
// ---------------------------------------------------------------------------

// GateRecord captures one gate-check evaluation for the audit log.
type GateRecord struct {
	Phase           int    `json:"phase"`
	Tier            string `json:"tier"`
	Action          string `json:"action"`
	ToolCalls       int    `json:"tool_calls"`
	Turns           int    `json:"turns"`
	IssuesCompleted int    `json:"issues_completed"`
}

// ---------------------------------------------------------------------------
// StateMachine
// ---------------------------------------------------------------------------

// StateMachine tracks the current orchestrator phase, runtime signals,
// and gate-check history.
type StateMachine struct {
	phase          int
	signals        capacity.Signals
	gateHistory    []GateRecord
	parallelCoders int
}

// New creates a StateMachine with the given parallel-coder count.
// If parallelCoders <= 0 it defaults to 5.
func New(parallelCoders int) *StateMachine {
	if parallelCoders <= 0 {
		parallelCoders = 5
	}
	return &StateMachine{
		phase:          0,
		signals:        capacity.Signals{ParallelCoders: parallelCoders},
		gateHistory:    nil,
		parallelCoders: parallelCoders,
	}
}

// Phase returns the current phase number.
func (sm *StateMachine) Phase() int { return sm.phase }

// Tier returns the current capacity tier based on accumulated signals.
func (sm *StateMachine) Tier() capacity.Tier {
	return capacity.ClassifyTier(sm.signals)
}

// Transition attempts to move from the current phase to target.
// It validates the transition, performs a gate check, and records the result.
// Returns the gate action or an error.
func (sm *StateMachine) Transition(target int) (capacity.Action, error) {
	// Validate transition.
	allowed, ok := validTransitions[sm.phase]
	if !ok || !allowed[target] {
		return "", &InvalidTransition{From: sm.phase, To: target}
	}

	// Gate check.
	tier := sm.Tier()
	action := capacity.GateAction(target, tier)

	// Record.
	sm.gateHistory = append(sm.gateHistory, GateRecord{
		Phase:           target,
		Tier:            string(tier),
		Action:          string(action),
		ToolCalls:       sm.signals.ToolCalls,
		Turns:           sm.signals.Turns,
		IssuesCompleted: sm.signals.IssuesCompleted,
	})

	// Shutdown actions.
	if action == capacity.Checkpoint || action == capacity.EmergencyStop {
		sm.phase = target
		return action, &ShutdownRequired{
			Tier:        tier,
			Action:      action,
			GatesPassed: len(sm.gateHistory),
			Signals:     sm.signals,
		}
	}

	sm.phase = target
	return action, nil
}

// RecordToolCall increments the tool-call counter and returns the new tier.
func (sm *StateMachine) RecordToolCall() capacity.Tier {
	sm.signals.ToolCalls++
	return sm.Tier()
}

// RecordTurn increments the turn counter and returns the new tier.
func (sm *StateMachine) RecordTurn() capacity.Tier {
	sm.signals.Turns++
	return sm.Tier()
}

// RecordIssueCompleted increments the issues-completed counter and returns the new tier.
func (sm *StateMachine) RecordIssueCompleted() capacity.Tier {
	sm.signals.IssuesCompleted++
	return sm.Tier()
}

// GateHistory returns a copy of the gate-check history.
func (sm *StateMachine) GateHistory() []GateRecord {
	out := make([]GateRecord, len(sm.gateHistory))
	copy(out, sm.gateHistory)
	return out
}

// ToDict serializes the state machine to a map for JSON persistence.
func (sm *StateMachine) ToDict() map[string]interface{} {
	history := make([]interface{}, len(sm.gateHistory))
	for i, r := range sm.gateHistory {
		history[i] = map[string]interface{}{
			"phase":            r.Phase,
			"tier":             r.Tier,
			"action":           r.Action,
			"tool_calls":       r.ToolCalls,
			"turns":            r.Turns,
			"issues_completed": r.IssuesCompleted,
		}
	}
	return map[string]interface{}{
		"phase":           sm.phase,
		"tool_calls":      sm.signals.ToolCalls,
		"turns":           sm.signals.Turns,
		"issues_completed": sm.signals.IssuesCompleted,
		"parallel_coders": sm.parallelCoders,
		"system_warning":  sm.signals.SystemWarning,
		"degraded_recall": sm.signals.DegradedRecall,
		"gate_history":    history,
	}
}

// FromDict restores a StateMachine from a serialized map (as produced by ToDict).
func FromDict(data map[string]interface{}) *StateMachine {
	sm := &StateMachine{}
	sm.phase = toInt(data["phase"])
	sm.signals.ToolCalls = toInt(data["tool_calls"])
	sm.signals.Turns = toInt(data["turns"])
	sm.signals.IssuesCompleted = toInt(data["issues_completed"])
	sm.parallelCoders = toInt(data["parallel_coders"])
	if sm.parallelCoders <= 0 {
		sm.parallelCoders = 5
	}
	sm.signals.ParallelCoders = sm.parallelCoders

	if sw, ok := data["system_warning"].(bool); ok {
		sm.signals.SystemWarning = sw
	}
	if dr, ok := data["degraded_recall"].(bool); ok {
		sm.signals.DegradedRecall = dr
	}

	if hist, ok := data["gate_history"].([]interface{}); ok {
		for _, entry := range hist {
			if m, ok := entry.(map[string]interface{}); ok {
				sm.gateHistory = append(sm.gateHistory, GateRecord{
					Phase:           toInt(m["phase"]),
					Tier:            fmt.Sprintf("%v", m["tier"]),
					Action:          fmt.Sprintf("%v", m["action"]),
					ToolCalls:       toInt(m["tool_calls"]),
					Turns:           toInt(m["turns"]),
					IssuesCompleted: toInt(m["issues_completed"]),
				})
			}
		}
	}

	return sm
}

// toInt converts an interface{} to int, handling both int and float64
// (the latter is common after JSON round-tripping).
func toInt(v interface{}) int {
	switch n := v.(type) {
	case int:
		return n
	case float64:
		return int(n)
	case int64:
		return int(n)
	default:
		return 0
	}
}
